File: LeetCode_Python/List&LinkList/21.merge-two-sorted-lists_.py
# @before-stub-for-debug-begin
from python3problem21 import *
from typing import *
# @before-stub-for-debug-end


# @lc app=leetcode id=21 lang=python3
#
# [21] Merge Two Sorted Lists
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(-1)
        p = dummy
        while ((list1 != None) and (list2 != None)):
            if (list1.val < list2.val):
                p.next = list1
                list1 = list1.next
            else:
                p.next = list2
                list2 = list2.next
            p = p.next
        if (list1 != None) :
            p.next = list1
        if (list2 != None) :
            p.next = list2
        return dummy.next

    # 2个指针p1, p2, 一个A头一个B头
    # 当指针不null, 比较2个指针
    # 比较小的一个插到新的链表并且把指针指向当前的下一个
    # O(n)

# @lc code=end

File: LeetCode_Go/List&LinkedList/21.merge-two-sorted-lists.go
/*
 * @lc app=leetcode id=21 lang=golang
 *
 * [21] Merge Two Sorted Lists
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// Merge two sorted linked lists and return it as a sorted list.

// traverse two lists; Two pointers with steps of 1, compare the values and append to the new list. when one pointer reaches the end, append the other list to the new list.

// O(n)

func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
	// dummy always point to the start of solution
	dummy := &ListNode{-1, nil}
	// tail point to the end of current solution
	tail := dummy

	// not nil list
	for list1 != nil && list2 != nil {

		if list1.Val < list2.Val { // 1 is smaller
			tail.Next = list1
			list1 = list1.Next
		} else { // 2 is smaller
			tail.Next = list2
			list2 = list2.Next
		}

		tail = tail.Next
	}

	if list1 == nil {
		tail.Next = list2
	}
	if list2 == nil {
		tail.Next = list1
	}

	return dummy.Next
}

// [1,2,4]\n[1,3,4]
// @lc code=end

